crates/network/src/lib.rs

```
@@ -175,16 +221,52 @@ impl TransactionBuilder<Optimism> for OpTransactionRequest {
     }
 
     fn prep_for_submission(&mut self) {
-        self.as_mut().prep_for_submission();
+        self.transaction_type = Some(self.preferred_type() as u8);
+        self.trim_conflicting_keys();
     }
 
     fn build_unsigned(self) -> BuildResult<OpTypedTransaction, Optimism> {
-        if let Err((tx_type, missing)) = self.as_ref().missing_keys() {
+        if let Err((tx_type, missing)) = self.missing_keys() {
             let tx_type = OpTxType::try_from(tx_type as u8).unwrap();
             return Err(TransactionBuilderError::InvalidTransactionRequest(tx_type, missing)
                 .into_unbuilt(self));
         }
-        Ok(self.build_typed_tx().expect("checked by missing_keys"))
+
+        let eth_typed = self.build_typed_tx().expect("checked by missing_keys");
+
+        match eth_typed {
+            EthereumTypedTransaction::Legacy(tx) => {
+                let tx = OpTypedTransaction::Legacy(tx);
+                Ok(tx)
+            }
+            EthereumTypedTransaction::Eip2930(tx) => {
+                let tx = OpTypedTransaction::Eip2930(tx);
+                Ok(tx)
+            }
+            EthereumTypedTransaction::Eip1559(tx) => {
+                let tx = OpTypedTransaction::Eip1559(tx);
+                Ok(tx)
+            }
+            EthereumTypedTransaction::Eip4844(tx) => {
+                // Converted to EIP-1559
+                let tx: TxEip4844 = tx.into();
+                Ok(OpTypedTransaction::Eip1559(TxEip1559 {
+                    chain_id: tx.chain_id,
+                    nonce: tx.nonce,
+                    gas_limit: tx.gas_limit,
+                    max_priority_fee_per_gas: tx.max_priority_fee_per_gas,
+                    max_fee_per_gas: tx.max_fee_per_gas,
+                    to: TxKind::Call(tx.to),
+                    value: tx.value,
+                    access_list: tx.access_list,
+                    input: tx.input,
+                }))
+            }
```

this doesn't seem right, this should return an error instead and possibly even panic in debug mode. eip4844 transactions shouldn't be accepted in any part of the op tx processing pipeline.

---

It's equivalent to `build_typed_tx` currently on main https://github.com/alloy-rs/op-alloy/blob/main/crates/rpc-types/src/transaction/request.rs#L104-L117 which doesn't return an error.

---

ok can be addressed in follow up